File: jre/src/main/java/com/azure/data/cosmos/serialization/hybridrow/layouts/LayoutBuilder.java
//------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
//------------------------------------------------------------

package com.azure.data.cosmos.serialization.hybridrow.layouts;

import com.azure.data.cosmos.serialization.hybridrow.SchemaId;
import com.azure.data.cosmos.serialization.hybridrow.schemas.StorageKind;

import java.util.ArrayList;
import java.util.Stack;

import static com.google.common.base.Preconditions.checkArgument;

public final class LayoutBuilder {
    private LayoutBit.Allocator bitallocator;
    private ArrayList<LayoutColumn> fixedColumns;
    private int fixedCount;
    private int fixedSize;
    private String name;
    private SchemaId schemaId = new SchemaId();
    private Stack<LayoutColumn> scope;
    private ArrayList<LayoutColumn> sparseColumns;
    private int sparseCount;
    private ArrayList<LayoutColumn> varColumns;
    private int varCount;

    // [ <present bits>
    //   <bool bits>
    //   <fixed_1> <fixed_2> ... <fixed_n>
    //   <var_1> <var_2> ... <var_n>
    //   <sparse_1> <sparse_2> ... <sparse_o>
    // ]
    public LayoutBuilder(String name, SchemaId schemaId) {
        this.name = name;
        this.schemaId = schemaId.clone();
        this.Reset();
    }

    public void AddFixedColumn(String path, LayoutType type, boolean nullable) {
        AddFixedColumn(path, type, nullable, 0);
    }

    //C# TO JAVA CONVERTER NOTE: Java does not support optional parameters. Overloaded method(s) are created above:
    //ORIGINAL LINE: public void AddFixedColumn(string path, LayoutType type, bool nullable, int length = 0)
    public void AddFixedColumn(String path, LayoutType type, boolean nullable, int length) {
        checkArgument(length >= 0);
        checkArgument(!type.getIsVarint());

        LayoutColumn col;
        if (type.getIsNull()) {
            checkArgument(nullable);
            LayoutBit nullbit = this.bitallocator.Allocate().clone();
            col = new LayoutColumn(path, type, TypeArgumentList.Empty, StorageKind.Fixed, this.getParent(),
                this.fixedCount, 0, nullbit.clone(), LayoutBit.Invalid, 0);
        } else if (type.getIsBool()) {
            LayoutBit nullbit = nullable ? this.bitallocator.Allocate() : LayoutBit.Invalid;
            LayoutBit boolbit = this.bitallocator.Allocate().clone();
            col = new LayoutColumn(path, type, TypeArgumentList.Empty, StorageKind.Fixed, this.getParent(),
                this.fixedCount, 0, nullbit.clone(), boolbit.clone(), 0);
        } else {
            LayoutBit nullBit = nullable ? this.bitallocator.Allocate() : LayoutBit.Invalid;
            col = new LayoutColumn(path, type, TypeArgumentList.Empty, StorageKind.Fixed, this.getParent(),
                this.fixedCount, this.fixedSize, nullBit.clone(), LayoutBit.Invalid, length);

            this.fixedSize += type.getIsFixed() ? type.Size : length;
        }

        this.fixedCount++;
        this.fixedColumns.add(col);
    }

    public void AddObjectScope(String path, LayoutType type) {
        LayoutColumn col = new LayoutColumn(path, type, TypeArgumentList.Empty, StorageKind.Sparse, this.getParent(),
            this.sparseCount, -1, LayoutBit.Invalid, LayoutBit.Invalid, 0);

        this.sparseCount++;
        this.sparseColumns.add(col);
        this.scope.push(col);
    }

    public void AddSparseColumn(String path, LayoutType type) {
        LayoutColumn col = new LayoutColumn(path, type, TypeArgumentList.Empty, StorageKind.Sparse, this.getParent(),
            this.sparseCount, -1, LayoutBit.Invalid, LayoutBit.Invalid, 0);

        this.sparseCount++;
        this.sparseColumns.add(col);
    }

    public void AddTypedScope(String path, LayoutType type, TypeArgumentList typeArgs) {
        LayoutColumn col = new LayoutColumn(path, type, typeArgs.clone(), StorageKind.Sparse, this.getParent(),
            this.sparseCount, -1, LayoutBit.Invalid, LayoutBit.Invalid, 0);

        this.sparseCount++;
        this.sparseColumns.add(col);
    }

    //C# TO JAVA CONVERTER NOTE: Java does not support optional parameters. Overloaded method(s) are created above:
    //ORIGINAL LINE: public void AddVariableColumn(string path, LayoutType type, int length = 0)
    public void AddVariableColumn(String path, LayoutType type, int length) {

        checkArgument(length >= 0);
        checkArgument(type.getAllowVariable());

        LayoutColumn col = new LayoutColumn(path, type, TypeArgumentList.Empty, StorageKind.Variable,
            this.getParent(), this.varCount, this.varCount, this.bitallocator.Allocate().clone(), LayoutBit.Invalid,
            length);

        this.varCount++;
        this.varColumns.add(col);
    }

    public void AddVariableColumn(String path, LayoutType type) {
        AddVariableColumn(path, type, 0);
    }

    public Layout Build() {
        // Compute offset deltas.  Offset bools by the present byte count, and fixed fields by the sum of the present
        // and bool count.
        int fixedDelta = this.bitallocator.getNumBytes();
        int varIndexDelta = this.fixedCount;

        // Update the fixedColumns with the delta before freezing them.
        ArrayList<LayoutColumn> updatedColumns =
            new ArrayList<LayoutColumn>(this.fixedColumns.size() + this.varColumns.size());

        for (LayoutColumn c : this.fixedColumns) {
            c.SetOffset(c.getOffset() + fixedDelta);
            updatedColumns.add(c);
        }

        for (LayoutColumn c : this.varColumns) {
            // Adjust variable column indexes such that they begin immediately following the last fixed column.
            c.SetIndex(c.getIndex() + varIndexDelta);
            updatedColumns.add(c);
        }

        updatedColumns.addAll(this.sparseColumns);

        Layout layout = new Layout(this.name, this.schemaId.clone(), this.bitallocator.getNumBytes(), this.fixedSize + fixedDelta, updatedColumns);
        this.Reset();
        return layout;
    }

    public void EndObjectScope() {
        checkArgument(this.scope.size() > 0);
        this.scope.pop();
    }

    private LayoutColumn getParent() {
        if (this.scope.empty()) {
            return null;
        }

        return this.scope.peek();
    }

    private void Reset() {
        this.bitallocator = new LayoutBit.Allocator();
        this.fixedSize = 0;
        this.fixedCount = 0;
        this.fixedColumns = new ArrayList<LayoutColumn>();
        this.varCount = 0;
        this.varColumns = new ArrayList<LayoutColumn>();
        this.sparseCount = 0;
        this.sparseColumns = new ArrayList<LayoutColumn>();
        this.scope = new Stack<LayoutColumn>();
    }
}

File: src/Serialization/HybridRow.Native/LayoutBuilder.cpp
// ------------------------------------------------------------
//  Copyright (c) Microsoft Corporation.  All rights reserved.
// ------------------------------------------------------------

#include "pch.h"
#include "LayoutBuilder.h"

#include "LayoutType.h"

// ReSharper disable IdentifierTypo
namespace cdb_hr
{
  LayoutBuilder::LayoutBuilder(std::string_view name, SchemaId schemaId) noexcept :
    m_name{name},
    m_schemaId{schemaId},
    m_fixedSize{},
    m_fixedCount{},
    m_varCount{},
    m_sparseCount{},
    m_bitAllocator{},
    m_fixedColumns{},
    m_varColumns{},
    m_sparseColumns{},
    m_scope{}
  {
    // [ <present bits>
    //   <bool bits>
    //   <fixed_1> <fixed_2> ... <fixed_n>
    //   <var_1> <var_2> ... <var_n>
    //   <sparse_1> <sparse_2> ... <sparse_o>
    // ]
  }

  LayoutColumn* LayoutBuilder::GetParent() noexcept
  {
    if (m_scope.empty())
    {
      return nullptr;
    }

    return m_scope.back();
  }

  void LayoutBuilder::AddFixedColumn(std::string_view path, const LayoutType* type, bool nullable,
                                     uint32_t length) noexcept
  {
    cdb_core::Contract::Requires(!type->IsVarint());

    std::unique_ptr<LayoutColumn> col;
    if (type->IsNull())
    {
      cdb_core::Contract::Requires(nullable);
      LayoutBit nullbit = m_bitAllocator.Allocate();
      col.reset(new LayoutColumn(
        path,
        type,
        {},
        StorageKind::Fixed,
        GetParent(),
        m_fixedCount,
        0,
        nullbit,
        LayoutBit::Invalid() // boolBit
      ));
    }
    else if (type->IsBool())
    {
      LayoutBit nullbit = nullable ? m_bitAllocator.Allocate() : LayoutBit::Invalid();
      LayoutBit boolbit = m_bitAllocator.Allocate();
      col.reset(new LayoutColumn(
        path,
        type,
        {},
        StorageKind::Fixed,
        GetParent(),
        m_fixedCount,
        0,
        nullbit,
        boolbit // boolBit
      ));
    }
    else
    {
      LayoutBit nullbit = nullable ? m_bitAllocator.Allocate() : LayoutBit::Invalid();
      col.reset(new LayoutColumn(
        path,
        type,
        {},
        StorageKind::Fixed,
        GetParent(),
        m_fixedCount,
        m_fixedSize,
        nullbit,
        LayoutBit::Invalid(), // boolBit
        length));

      m_fixedSize += type->IsFixed() ? type->GetSize() : length;
    }

    m_fixedCount++;
    cdb_core::Contract::Assert(col != nullptr);
    m_fixedColumns.push_back(std::move(col));
  }

  void LayoutBuilder::AddVariableColumn(std::string_view path, const LayoutType* type, uint32_t length) noexcept
  {
    cdb_core::Contract::Requires(type->AllowVariable());

    std::unique_ptr<LayoutColumn> col{
      new LayoutColumn{
        path,
        type,
        {},
        StorageKind::Variable,
        GetParent(),
        m_varCount,
        m_varCount,
        m_bitAllocator.Allocate(), // nullBit
        LayoutBit::Invalid(), // boolBit
        length
      }
    };

    m_varCount++;
    m_varColumns.push_back(std::move(col));
  }

  void LayoutBuilder::AddSparseColumn(std::string_view path, const LayoutType* type) noexcept
  {
    std::unique_ptr<LayoutColumn> col{
      new LayoutColumn{
        path,
        type,
        {},
        StorageKind::Sparse,
        GetParent(),
        m_sparseCount,
        UINT_MAX,
        LayoutBit::Invalid(), // nullBit
        LayoutBit::Invalid() // boolBit
      }
    };

    m_sparseCount++;
    m_sparseColumns.push_back(std::move(col));
  }

  void LayoutBuilder::AddObjectScope(std::string_view path, const LayoutType* type) noexcept
  {
    std::unique_ptr<LayoutColumn> col{
      new LayoutColumn{
        path,
        type,
        {},
        StorageKind::Sparse,
        GetParent(),
        m_sparseCount,
        UINT_MAX,
        LayoutBit::Invalid(), // nullBit
        LayoutBit::Invalid() // boolBit
      }
    };

    m_sparseCount++;
    m_scope.push_back(col.get());
    m_sparseColumns.push_back(std::move(col));
  }

  void LayoutBuilder::EndObjectScope() noexcept
  {
    cdb_core::Contract::Requires(!m_scope.empty());
    m_scope.pop_back();
  }

  void LayoutBuilder::AddTypedScope(std::string_view path, const LayoutType* type, TypeArgumentList typeArgs) noexcept
  {
    std::unique_ptr<LayoutColumn> col{
      new LayoutColumn{
        path,
        type,
        std::move(typeArgs),
        StorageKind::Sparse,
        GetParent(),
        m_sparseCount,
        UINT_MAX,
        LayoutBit::Invalid(), // nullBit
        LayoutBit::Invalid() // boolBit
      }
    };

    m_sparseCount++;
    m_sparseColumns.push_back(std::move(col));
  }

  std::unique_ptr<Layout> LayoutBuilder::Build() noexcept
  {
    // Compute offset deltas.  Offset bools by the present byte count, and fixed fields by the sum of the present and bool count.
    uint32_t fixedDelta = m_bitAllocator.GetNumBytes();
    uint32_t varIndexDelta = m_fixedCount;

    // Update the fixedColumns with the delta before freezing them.
    tla::vector<std::unique_ptr<LayoutColumn>> updatedColumns{};
    updatedColumns.reserve(m_fixedColumns.size() + m_varColumns.size());

    for (auto& c : m_fixedColumns)
    {
      c->SetOffset(c->GetOffset() + fixedDelta);
      updatedColumns.push_back(std::move(c));
    }

    for (auto& c : m_varColumns)
    {
      // Adjust variable column indexes such that they begin immediately following the last fixed column.
      c->SetIndex(c->GetIndex() + varIndexDelta);
      updatedColumns.push_back(std::move(c));
    }

    for (auto& c : m_sparseColumns)
    {
      updatedColumns.push_back(std::move(c));
    }

    std::unique_ptr<Layout> layout = std::unique_ptr<Layout>{
      new Layout(m_name, m_schemaId, m_bitAllocator.GetNumBytes(), m_fixedSize + fixedDelta, std::move(updatedColumns))
    };
    Reset();
    return std::move(layout);
  }

  void LayoutBuilder::Reset() noexcept
  {
    m_bitAllocator = LayoutBit::Allocator();
    m_fixedSize = 0;
    m_fixedCount = 0;
    m_fixedColumns.clear();
    m_varCount = 0;
    m_varColumns.clear();
    m_sparseCount = 0;
    m_sparseColumns.clear();
    m_scope.clear();
  }
}
